llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp

```diff
@@ -1328,20 +1327,10 @@ Value *HWAddressSanitizer::tagPointer(IRBuilder<> &IRB, Type *Ty,
 
 // Remove tag from an address.
 Value *HWAddressSanitizer::untagPointer(IRBuilder<> &IRB, Value *PtrLong) {
+  unsigned SignExtShift = 64 - PointerTagShift;
   assert(!UsePageAliases);
-  Value *UntaggedPtrLong;
-  if (CompileKernel) {
-    // Kernel addresses have 0xFF in the most significant byte.
-    UntaggedPtrLong =
-        IRB.CreateOr(PtrLong, ConstantInt::get(PtrLong->getType(),
-                                               TagMaskByte << PointerTagShift));
```

Comment:
X86's TagMaskByte is 0x3F(6 bits) even though PointerTagShift is 57. Does that mean the original code doesn't set bit 63?

Reply:
Correct. X86 LAM requires that bit 63 == bit 56, so bit 63 is not available as a tag bit.

Reply:
If I read this correctly, new code does not preserve top 1s in kernel pointer?

Reply:
Yes, for x86, it will copy bit 56 to bit 63. Is this a problem?

Reply:
Looks fine for canonical addresses https://godbolt.org/z/z3rYjjbrz

Reply:
@ramosian-glider WDYT?